File: backend/src/main/java/ch/zuehlke/hatch/sailingserver/signalk/WebsocketClientEndpoint.java
package ch.zuehlke.hatch.sailingserver.signalk;

import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalKSubscibtionFactory;
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalKSubscriptionPath;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.client.ReactorNettyWebSocketClient;
import org.springframework.web.reactive.socket.client.WebSocketClient;
import reactor.core.publisher.Mono;

import java.net.URI;


@Component
public class WebsocketClientEndpoint {

    private final String SIGNALK_HOST = "localhost";
    private final String SIGNALK_PORT = "3000";

    private static final Logger log = LoggerFactory.getLogger(WebsocketClientEndpoint.class);

    public void connect() {
        WebSocketClient client = new ReactorNettyWebSocketClient();
        client.execute(
                URI.create("ws://" + SIGNALK_HOST + ":" + SIGNALK_PORT + "/signalk/v1/stream/?subscribe=none"),
                session ->
                        session
                                .send(Mono.just(session.textMessage(getInitialSubscription())))
                                .thenMany(session.receive()
                                        .map(WebSocketMessage::getPayloadAsText)
                                        .log())
                                .then())
                .subscribe();

    }

    private String getInitialSubscription() {
        return SignalKSubscibtionFactory.createDefaultSubscriptionWithSinglePath(
                    "*",
                    1000,
                    SignalKSubscriptionPath.FORMAT_DELTA,
                    SignalKSubscriptionPath.POLICY_INSTANT,
                    1000).toString();
    }

}

File: backend/src/main/kotlin/ch/zuehlke/hatch/sailingserver/signalk/WebsocketClientEndpoint.kt
package ch.zuehlke.hatch.sailingserver.signalk

import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalkSubscription
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SubscriptionInfo
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.client.ReactorNettyWebSocketClient
import reactor.core.publisher.Mono
import java.net.URI


@Component
class WebsocketClientEndpoint {

    @Value("\${signalk.subscription.endpoint}")
    lateinit var signalkSubscriptionEndpoint: String

    private val objectMapper = ObjectMapper()
    private val initialSubscription = SignalkSubscription(
            "vessels.self",
            listOf(SubscriptionInfo("*", "1000", "delta", "instant", "1000")))


    fun connect() {
        val client = ReactorNettyWebSocketClient()
        client.execute(
                URI.create(signalkSubscriptionEndpoint)
        ) { session ->
            session
                    .send(Mono.just(session.textMessage(objectMapper.writeValueAsString(initialSubscription))))
                    .thenMany(session.receive()
                            .map<String> { message -> message.payloadAsText }
                            .log())
                    .then()
        }
                .subscribe()

    }

}
